## 문서 군집화 개념
문서 군집화 (Document clustering)은 비슷한 텍스트 구성의 문서를 같은 카테고리 소속으로 분류하는 군집화 기법이다.    
텍스트 분류 기반의 문서 분류는 사전에 결정 카테고리 값을 가진 학습 데이터 세트가 필요하지만,    
문서 군집화는 학습 데이터 세트가 필요하지 않는 비지도 학습 기반으로 동작한다.    

리뷰 데이터 세트를 이용하여 문서 군집화를 수행하겠다.    
(해당 데이터셋 다운링크: https://archive.ics.uci.edu/dataset/191/opinosis+opinion+frasl+review    )

In [2]:
import pandas as pd
import glob, os
import warnings 
warnings.filterwarnings('ignore')
pd.set_option('display.max_colwidth', 700)

# 아래는 제 컴퓨터에서 압축 파일을 풀어 놓은 디렉터리이니, 여러분의 디렉터리를 설정해 주세요  
path = r'C:\Users\user\Desktop\AIstudy\MachineLearning\opinosis+opinion+frasl+review\OpinosisDataset1.0\topics'
# path로 지정한 디렉터리 밑에 있는 모든 .data 파일들의 파일명을 리스트로 취합
all_files = glob.glob(os.path.join(path, "*.data"))    
filename_list = []
opinion_text = []

# 개별 파일들의 파일명은 filename_list 리스트로 취합, 
# 개별 파일들의 파일 내용은 DataFrame 로딩 후 다시 string으로 변환하여 opinion_text 리스트로 취합 
for file_ in all_files:
    # 개별 파일을 읽어서 DataFrame으로 생성 
    df = pd.read_table(file_,index_col=None, header=0,encoding='latin1')
    
    # 절대경로로 주어진 file 명을 가공. 만일 Linux에서 수행시에는 아래 \\를 / 변경. 
    # 맨 마지막 .data 확장자도 제거
    filename_ = file_.split('\\')[-1]
    filename = filename_.split('.')[0]

    # 파일명 리스트와 파일 내용 리스트에 파일명과 파일 내용을 추가. 
    filename_list.append(filename)
    opinion_text.append(df.to_string())

# 파일명 리스트와 파일 내용 리스트를  DataFrame으로 생성
document_df = pd.DataFrame({'filename':filename_list, 'opinion_text':opinion_text})
document_df.head()

filename  \
0   accuracy_garmin_nuvi_255W_gps   
1  bathroom_bestwestern_hotel_sfo   
2      battery-life_amazon_kindle   
3      battery-life_ipod_nano_8gb   
4     battery-life_netbook_1005ha   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  opinion_text  
0                                                                                                                                                                                                                                                                 , and is very, very accurate .\n0                                                                                                                                                                           but for the most part, we find that the Garmin software provides accurate directions, whereever we intend to go .\n1                                                                                                              This functi...  
1                                                                                                                                                                                                                                                The room was not overly big, but clean and very comfortable beds, a great shower and very clean bathrooms .\n0                                                                                                                                                                                                                          The second room was smaller, with a very inconvenient bathroom layout, but at least it was quieter and we were able to sleep .\n1 ...  
2                                                                                                                                                                                                               After I plugged it in to my USB hub on my computer to charge the battery the charging cord design is very clever !\n0                                                                                                                                     After you have paged tru a 500, page book one, page, at, a, time to get from Chapter 2 to Chapter 15, see how excited you are about a low battery and all the time it took to get there !\n1                                                     ...  
3                                                                                                                                                                                                                                                                                                 short battery life  I moved up from an 8gb .\n0                                                                                                                                                                                                                                                                                            I love this ipod except for the battery life .\n1                             ...  
4                                                                                                                                                                                                                                                                                    

위의 문서 데이터를 TF-IDF 형태로 피처 벡터화를 진행하고 어근 변환을 위해 LemNormalize()함수를 생성    
(TfidfVectorizer는 Lemmatization 같은 어근 변환을 직접 지원하지 않기 때문에 만들어준다.)    

In [9]:
from nltk.stem import WordNetLemmatizer
import nltk
import string
nltk.download('all')

remove_punct_dict = dict((ord(punct), None) for punct in string.punctuation)
lemmar = WordNetLemmatizer()

# 입력으로 들어온 token단어들에 대해서 lemmatization 어근 변환. 
def LemTokens(tokens):
    return [lemmar.lemmatize(token) for token in tokens]

# TfidfVectorizer 객체 생성 시 tokenizer인자로 해당 함수를 설정하여 lemmatization 적용
# 입력으로 문장을 받아서 stop words 제거-> 소문자 변환 -> 단어 토큰화 -> lemmatization 어근 변환. 
def LemNormalize(text):
    return LemTokens(nltk.word_tokenize(text.lower().translate(remove_punct_dict)))

[nltk_data] Downloading collection 'all'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to
[nltk_data]    |     C:\Users\user\AppData\Roaming\nltk_data...
[nltk_data]    |   Package abc is already up-to-date!
[nltk_data]    | Downloading package alpino to
[nltk_data]    |     C:\Users\user\AppData\Roaming\nltk_data...
[nltk_data]    |   Package alpino is already up-to-date!
[nltk_data]    | Downloading package averaged_perceptron_tagger to
[nltk_data]    |     C:\Users\user\AppData\Roaming\nltk_data...
[nltk_data]    |   Package averaged_perceptron_tagger is already up-
[nltk_data]    |       to-date!
[nltk_data]    | Downloading package averaged_perceptron_tagger_ru to
[nltk_data]    |     C:\Users\user\AppData\Roaming\nltk_data...
[nltk_data]    |   Package averaged_perceptron_tagger_ru is already
[nltk_data]    |       up-to-date!
[nltk_data]    | Downloading package basque_grammars to
[nltk_data]    |     C:\Users\user\AppData\Roaming\nltk_data...
[nltk_data]    |   Pac

In [10]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vect = TfidfVectorizer(tokenizer=LemNormalize, stop_words='english' , \
                             ngram_range=(1,2), min_df=0.05, max_df=0.85 )

#opinion_text 컬럼값으로 feature vectorization 수행
feature_vect = tfidf_vect.fit_transform(document_df['opinion_text'])

피처의 갯수를 제한하고 개별 문서 텍스트에 대해 TF-IDF변환된 피처 벡터화 행렬을 구할 수 있다.

In [11]:
from sklearn.cluster import KMeans

# 5개 집합으로 군집화 수행. 예제를 위해 동일한 클러스터링 결과 도출용 random_state=0 
km_cluster = KMeans(n_clusters=5, max_iter=10000, random_state=0)
km_cluster.fit(feature_vect)
cluster_label = km_cluster.labels_
cluster_centers = km_cluster.cluster_centers_

군집화의 기법으로 K-평균을 사용하여 Centroid(중심)기반으로 군집화 진행    
document_df DataFrame에 cluster_label 칼럼을 추가해 저장한다.

In [12]:
# 군집이 각 주제별로 유사한 형태로 잘 구성되었는지 확인
document_df['cluster_label'] = cluster_label
document_df.head()

filename  \
0   accuracy_garmin_nuvi_255W_gps   
1  bathroom_bestwestern_hotel_sfo   
2      battery-life_amazon_kindle   
3      battery-life_ipod_nano_8gb   
4     battery-life_netbook_1005ha   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  opinion_text  \
0                                                                                                                                                                                                                                                                 , and is very, very accurate .\n0                                                                                                                                                                           but for the most part, we find that the Garmin software provides accurate directions, whereever we intend to go .\n1                                                                                                              This functi...   
1                                                                                                                                                                                                                                                The room was not overly big, but clean and very comfortable beds, a great shower and very clean bathrooms .\n0                                                                                                                                                                                                                          The second room was smaller, with a very inconvenient bathroom layout, but at least it was quieter and we were able to sleep .\n1 ...   
2                                                                                                                                                                                                               After I plugged it in to my USB hub on my computer to charge the battery the charging cord design is very clever !\n0                                                                                                                                     After you have paged tru a 500, page book one, page, at, a, time to get from Chapter 2 to Chapter 15, see how excited you are about a low battery and all the time it took to get there !\n1                                                     ...   
3                                                                                                                                                                                                                                                                                                 short battery life  I moved up from an 8gb .\n0                                                                                                                                                                                                                                                                                            I love this ipod except for the battery life .\n1                             ...   
4                                                                                                                                                                                                                                                                               

In [13]:
# sort_values(칼럼명)을 통해 데이터 정렬
document_df[document_df['cluster_label']==0].sort_values(by='filename')

filename  \
17  interior_honda_accord_2008   
18  interior_toyota_camry_2007   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
17                                                                                                                                                                                                                                                                                                                                                                                                                                                                             I love the new body style and the interior is a simple pleasure except for the center dash .\n0                                                                                                                                              ...   
18                                                                                                                                                                                                                                      First of all, the interior has way too many cheap plastic parts like the cheap plastic center piece that houses the clock .\n0                                                                                                                                                                       3 blown struts at 30,000 miles, interior trim coming loose and rattling squeaking, stains on paint, and bug splats taking paint off, premature uneven brake wear, on 3rd windsh...   

    cluster_label  
17              0  
18              0

Cluster #0은 호텔에 대한 내용이 군집화되어 있는 걸 확인할 수 있다.

In [14]:
document_df[document_df['cluster_label']==1].sort_values(by='filename')

filename  \
2    battery-life_amazon_kindle   
3    battery-life_ipod_nano_8gb   
4   battery-life_netbook_1005ha   
19      keyboard_netbook_1005ha   
26   performance_netbook_1005ha   
35         screen_ipod_nano_8gb   
36        screen_netbook_1005ha   
41     size_asus_netbook_1005ha   
42          sound_ipod_nano_8gb   
44               speed_windows7   
49          video_ipod_nano_8gb   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
2                                                                                                                                                                                                                After I plugged it in to my USB hub on my computer to charge the battery the charging cord design is very clever !\n0                                                                                                                                     After you have paged tru a 500, page book one, page, at, a, time to get from Chapter 2 to Chapter 15, see how excited you are about a low battery and all the time it took to get there !\n1                                                     ...   
3                                                                                                                                                                                                                                                                                                  short battery life  I moved up from an 8gb .\n0                                                                                                                                                                                                                                                                                            I love this ipod except for the battery life .\n1                             ...   
4                                                                                                                                                                                                                                                                                                                   6GHz 533FSB cpu, glossy display, 3, Cell 23Wh Li, ion Battery  , and a 1 .\n0                                                                                                                                                                                                                                                                                              Not to mention that as of now...   
19                                                                                                                                                                                  ,  I think the new keyboard rivals the great hp mini keyboards .\n0                                                                                                                                 Since the battery life difference is minimum, the only reason to upgrade would be to get the better keyboard .\n1                                                                                                                                                                                   The keyboard is now as good as t...   
26                                                                 

Cluster #1은 주로 차량용내비게이션에 대한 내용이 군집화되어 있는 걸 확인할 수 있다.    

중심개수를 5개에서 3개로 낮추어 결과를 확인해보자.

In [15]:
from sklearn.cluster import KMeans

# 3개의 집합으로 군집화 
km_cluster = KMeans(n_clusters=3, max_iter=10000, random_state=0)
km_cluster.fit(feature_vect)
cluster_label = km_cluster.labels_


# 소속 클러스터를 cluster_label 컬럼으로 할당하고 cluster_label 값으로 정렬
document_df['cluster_label'] = cluster_label
document_df.sort_values(by='cluster_label')

filename  \
24    parking_bestwestern_hotel_sfo   
21      location_holiday_inn_london   
32          room_holiday_inn_london   
15       free_bestwestern_hotel_sfo   
14           food_swissotel_chicago   
13          food_holiday_inn_london   
40  service_swissotel_hotel_chicago   
38    service_bestwestern_hotel_sfo   
20   location_bestwestern_hotel_sfo   
39       service_holiday_inn_london   
28         price_holiday_inn_london   
45      staff_bestwestern_hotel_sfo   
46          staff_swissotel_chicago   
30      rooms_bestwestern_hotel_sfo   
31          rooms_swissotel_chicago   
1    bathroom_bestwestern_hotel_sfo   
22        mileage_honda_accord_2008   
29        quality_toyota_camry_2007   
37          seats_honda_accord_2008   
25    performance_honda_accord_2008   
16    gas_mileage_toyota_camry_2007   
17       interior_honda_accord_2008   
7         comfort_toyota_camry_2007   
6         comfort_honda_accord_2008   
47   transmission_toyota_camry_2007   
18       interior_toyota_camry_2007   
41         size_asus_netbook_1005ha   
42              sound_ipod_nano_8gb   
43       speed_garmin_nuvi_255W_gps   
44                   speed_windows7   
36            screen_netbook_1005ha   
35             screen_ipod_nano_8gb   
48     updates_garmin_nuvi_255W_gps   
34      screen_garmin_nuvi_255W_gps   
0     accuracy_garmin_nuvi_255W_gps   
27              price_amazon_kindle   
26       performance_netbook_1005ha   
49              video_ipod_nano_8gb   
23         navigation_amazon_kindle   
19          keyboard_netbook_1005ha   
12              fonts_amazon_kindle   
11                features_windows7   
10    eyesight-issues_amazon_kindle   
9      display_garmin_nuvi_255W_gps   
8   directions_garmin_nuvi_255W_gps   
5             buttons_amazon_kindle   
4       battery-life_netbook_1005ha   
3        battery-life_ipod_nano_8gb   
2        battery-life_amazon_kindle   
33   satellite_garmin_nuvi_255W_gps   
50       voice_garmin_nuvi_255W_gps   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   opinion_text  \
24                                                                                                                                                                                                                       Parking was expensive but I think this is common for San Fran .\n0                                                                                                                                                                                        there is a fee for parking but well worth it seeing no where to park if you do have a car .\n1                                                                                                                                           ...   
21                                                                                                                                                                                                                                                                                                                  Great location for tube and we crammed in a fair amount of sightseeing in a short time .\n0                                                                                                                                                                                         

Cluster #0은 포터블 전자기기, #1은 자동차 리뷰, #2는 호텔 리뷰로 군집이 된 것을 확인할 수 있다.    

## 군집별 핵심단어 추출하기
핵심단어를 기준으로 군집화를 했으니 이번에는 각 군집을 구성하는 핵심단어를 확인해보자.    

In [16]:
cluster_centers = km_cluster.cluster_centers_
print('cluster_centers shape :',cluster_centers.shape)
print(cluster_centers)

cluster_centers shape : (3, 4611)
[[0.         0.00099499 0.00174637 ... 0.         0.00183397 0.00144581]
 [0.         0.00092551 0.         ... 0.         0.         0.        ]
 [0.01005322 0.         0.         ... 0.00706287 0.         0.        ]]


cluster_centers_ 속성은 각 군집을 구성하는 단어 피처가 군집 중심(centroid)을 기준으로 얼마나 가깝게 위치해 있는지 알 수 있다.    
cluster_centers_가 3,4611 배열로 군집이 3개 피처가 4611로 구성되었음을 알 수 있으며 각 피처가 중심값으로부터 얼마나 가까운 값을 가지고 있는지 상대적인 값이 출력된 것이다.    
(1에 가까울수록 중심값과 가까운 값을 의미)

In [17]:

# 군집별 top n 핵심단어, 그 단어의 중심 위치 상대값, 대상 파일명들을 반환함. 
def get_cluster_details(cluster_model, cluster_data, feature_names, clusters_num, top_n_features=10):
    cluster_details = {}
    
    # cluster_centers array 의 값이 큰 순으로 정렬된 index 값을 반환
    # 군집 중심점(centroid)별 할당된 word 피처들의 거리값이 큰 순으로 값을 구하기 위함.  
    centroid_feature_ordered_ind = cluster_model.cluster_centers_.argsort()[:,::-1]
    
    #개별 군집별로 iteration하면서 핵심단어, 그 단어의 중심 위치 상대값, 대상 파일명 입력
    for cluster_num in range(clusters_num):
        # 개별 군집별 정보를 담을 데이터 초기화. 
        cluster_details[cluster_num] = {}
        cluster_details[cluster_num]['cluster'] = cluster_num
        
        # cluster_centers_.argsort()[:,::-1] 로 구한 index 를 이용하여 top n 피처 단어를 구함. 
        top_feature_indexes = centroid_feature_ordered_ind[cluster_num, :top_n_features]
        top_features = [ feature_names[ind] for ind in top_feature_indexes ]
        
        # top_feature_indexes를 이용해 해당 피처 단어의 중심 위치 상댓값 구함 
        top_feature_values = cluster_model.cluster_centers_[cluster_num, top_feature_indexes].tolist()
        
        # cluster_details 딕셔너리 객체에 개별 군집별 핵심 단어와 중심위치 상대값, 그리고 해당 파일명 입력
        cluster_details[cluster_num]['top_features'] = top_features
        cluster_details[cluster_num]['top_features_value'] = top_feature_values
        filenames = cluster_data[cluster_data['cluster_label'] == cluster_num]['filename']
        filenames = filenames.values.tolist()
        cluster_details[cluster_num]['filenames'] = filenames
        
    return cluster_details

In [18]:

def print_cluster_details(cluster_details):
    for cluster_num, cluster_detail in cluster_details.items():
        print('####### Cluster {0}'.format(cluster_num))
        print('Top features:', cluster_detail['top_features'])
        print('Reviews 파일명 :',cluster_detail['filenames'][:7])
        print('==================================================')

In [20]:
feature_names = tfidf_vect.get_feature_names_out()

cluster_details = get_cluster_details(cluster_model=km_cluster, cluster_data=document_df,\
                                  feature_names=feature_names, clusters_num=3, top_n_features=10 )
print_cluster_details(cluster_details)

####### Cluster 0
Top features: ['room', 'hotel', 'service', 'staff', 'food', 'location', 'bathroom', 'clean', 'price', 'parking']
Reviews 파일명 : ['bathroom_bestwestern_hotel_sfo', 'food_holiday_inn_london', 'food_swissotel_chicago', 'free_bestwestern_hotel_sfo', 'location_bestwestern_hotel_sfo', 'location_holiday_inn_london', 'parking_bestwestern_hotel_sfo']
####### Cluster 1
Top features: ['interior', 'seat', 'mileage', 'comfortable', 'gas', 'gas mileage', 'transmission', 'car', 'performance', 'quality']
Reviews 파일명 : ['comfort_honda_accord_2008', 'comfort_toyota_camry_2007', 'gas_mileage_toyota_camry_2007', 'interior_honda_accord_2008', 'interior_toyota_camry_2007', 'mileage_honda_accord_2008', 'performance_honda_accord_2008']
####### Cluster 2
Top features: ['screen', 'battery', 'keyboard', 'battery life', 'life', 'kindle', 'direction', 'video', 'size', 'voice']
Reviews 파일명 : ['accuracy_garmin_nuvi_255W_gps', 'battery-life_amazon_kindle', 'battery-life_ipod_nano_8gb', 'battery-life_